### DISCUSSION TITLE
Pip dependencies not getting cached even after successful restore

### DISCUSSION THREAD

**Discussion by sayakpaul** [community/community#26096] [Actions]
<p>Hi folks.</p>
<p>I am new to GitHub Actions so apologies if I am asking a naive question.</p>
<p>I am trying to set up my workflow such that after the pip dependencies get installed for the first time they get cached and do not get installed next time when the workflow is triggered.</p>
<p>My workflow file looks like so:</p>
<pre><code class="lang-auto">name: Metric logging with wandb
on: [push]
jobs:
  run:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - name: Set up Python 3.x
        uses: actions/setup-python@v2
        with:
          python-version: '3.7' 
      - uses: actions/cache@master
        id: cache
        with:
          path: .cache/pip
          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
          restore-keys: ${{ runner.os }}-test
      - name: Install Dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: |
          echo "Installing dependencies and caching them."
          pip install -r requirements.txt
      - name: Train model and log wandb metric
        run: |
          # Your ML workflow goes here
          python train.py --key ${{ secrets.WANDB_API_KEY }}
          echo "Finished run."
</code></pre>
<p>The second time the workflow is getting trigger there is a cache hit and it is able to restore the cache as well:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/283d016a74ffec111a596cfe10d7e2ca8f26f671.png" data-download-href="/uploads/short-url/5JXNL3C4yqB2RkMGYUTnelONjhf.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/22957388/181095203-bc833c0f-e918-4c88-b640-942478c13e40.png" alt="image" data-base62-sha1="5JXNL3C4yqB2RkMGYUTnelONjhf" width="690" height="140" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/2/283d016a74ffec111a596cfe10d7e2ca8f26f671_2_690x140.png, https://aws1.discourse-cdn.com/github/optimized/2X/2/283d016a74ffec111a596cfe10d7e2ca8f26f671_2_1035x210.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/2/283d016a74ffec111a596cfe10d7e2ca8f26f671.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/283d016a74ffec111a596cfe10d7e2ca8f26f671_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1313×268 40.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>But then again I get the following:</p>
<p><img src="https://user-images.githubusercontent.com/22957388/181095205-c3e91662-3b4e-4040-a320-ca5cbcb30b9f.png" alt="image" data-base62-sha1="h40ZsgqdAYdrIsme66919LfkhBB" width="372" height="147"></p>
<p><code>pandas</code> is not a new requirement that was already installed in the previous run.</p>
<p>Am I missing out on something?</p>

**Comment by sayakpaul** [comment]
<p>I modified my workflow file a bit like the following:</p>
<pre><code class="lang-auto">name: Metric logging with wandb

on: [push]

jobs:
  run:

    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
      - name: Set up Python 3.x
        uses: actions/setup-python@v2
        with:
          python-version: '3.7' 
      - uses: actions/cache@master
        id: cache
        with:
          path: ~/.cache/pip
          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
          restore-keys: |
            ${{ runner.os }}-pip-
            ${{ runner.os }}-
      - name: Train model and log wandb metric
        run: |
          # Your ML workflow goes here
          pip install -r requirements.txt
          python train.py --key ${{ secrets.WANDB_API_KEY }}
          echo "Finished run."
</code></pre>
<p>But still the dependencies install and do not get cached.</p>

**Comment by NicoNBW** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a><br>
Do you have a link to the public repository or a minimal example repository that shows this behaviour?<br>
By the way, could you please add <code>ls ~/.cache/pip/</code> before “<code>Train model and log wandb metric</code>” step to check if <code>pandas</code> has been installed in the previous run?</p>

**Comment by sayakpaul** [comment]
<p><span class="mention">@niconbwthanks</span> for your reply. Here’e the branch where I tested this: <a href="https://github.com/sayakpaul/wine/tree/wandb-integration./" rel="nofollow noopener">https://github.com/sayakpaul/wine/tree/wandb-integration./</a></p>

**Comment by NicoNBW** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a><br>
The reason for this problem should be that the specified cache path is incorrect. You could try to specified cache path, such as:</p>
<pre><code class="lang-auto">      - uses: actions/cache@master
        id: cache
        with:
          path: /opt/hostedtoolcache/Python/3.7.8/x64/lib/python3.7/site-packages/*
          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
          restore-keys: |
            ${{ runner.os }}-pip-
            ${{ runner.os }}-
</code></pre>
<p>More about <a href="https://docs.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions#specifying-a-python-version" rel="nofollow noopener">locations for the tools cache in each GitHub-hosted runner</a>.<br>
By the way, there is an environment variable called <a href="https://github.com/marketplace/actions/setup-python#hosted-tool-cache" rel="nofollow noopener">RUNNER_TOOL_CACHE</a> on each GitHub hosted runner that describes the location of this tools cache and there is where you will find Python and PyPy installed. In your scenario, <code>pandas</code> should has been installed in the following path:<br>
<code>$RUNNER_TOOL_CACHE/Python/3.7.8/x64/lib/python3.7/site-packages/pandas</code></p>

**Comment by sayakpaul** [comment]
<p>Thank you for the pointers, <a class="mention" href="/niconbw">@niconbw</a>. Could you also elaborate as to why the message “Cache restored …” was there in my logs? Am I missing out on something?</p>

**Comment by NicoNBW** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a><br>
We can find that you are using the <code>~/.cache/pip</code> path in the cache@master action, but <code>pandas</code>  has not been installed under this path. Even if you save the cache under this path(<code>~/.cache/pip</code>), you still cannot find the module (such as <code>pandas</code>) when you restore the cache successfully.</p>

**Comment by sayakpaul** [comment]
<p>Fair enough <a class="mention" href="/niconbw">@niconbw</a>. Thank you for your guidance.</p>

**Comment by sayakpaul** [comment]
<p><a class="mention" href="/niconbw">@niconbw</a> I tried your suggestions but I am still not sure if caching is indeed being reflected. Here are the logs: <a href="https://github.com/sayakpaul/wine/runs/975167639?check_suite_focus=true" rel="nofollow noopener">https://github.com/sayakpaul/wine/runs/975167639?check_suite_focus=true</a>.</p>

**Comment by NicoNBW** [comment, marked as answer]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a><br>
Please try again with the following yml:</p>
<pre><code class="lang-auto">jobs:
  run:
    runs-on: ubuntu-latest
    
    steps:
      - uses: actions/checkout@v2
      - name: Set up Python 3.x
        uses: actions/setup-python@v2
        with:
          python-version: '3.7'    
       
      - uses: actions/cache@master
        id: cache
        with:
          path: ${{ env.pythonLocation}}/lib/python3.7/site-packages/*
          key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
          restore-keys: |
            ${{ runner.os }}-pip-
            ${{ runner.os }}-
            
      - name: Install Dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: |
          echo "Installing dependencies and caching them."
          pip install -r requirements.txt
     
      - name: ls show env.pythonLocation
        run: ls -R ${{ env.pythonLocation}}
     
      - name: show module
        run: pip show pandas
      
      - name: Train model and log wandb metric
        run:  echo "Finished run."
</code></pre>

**Comment by sayakpaul** [comment]
<p><a class="mention" href="/niconbw">@niconbw</a> you probably forgot to include <code>python train.py</code>.</p>

**Comment by NicoNBW** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a><br>
sorry. Please add <code>python train.py</code> back in your side.</p>

**Comment by sayakpaul** [comment]
<p><a class="mention" href="/niconbw">@niconbw</a> I appended the necessary stuff in the YAML file and I got it to working:<br>
<a href="https://github.com/sayakpaul/wine/runs/979690608?check_suite_focus=true" rel="nofollow noopener">https://github.com/sayakpaul/wine/runs/979690608?check_suite_focus=true</a>.</p>
<p>Thanks much.</p>

**Comment by NicoNBW** [comment]
<p>Hi <a class="mention" href="/sayakpaul">@sayakpaul</a><br>
You’re welcome. It’s my pleasure to help you.</p>

**Comment by Djpod** [comment]
<p>jobs:<br>
run:<br>
runs-on: ubuntu-latest</p>
<pre><code>steps:
  - uses: actions/checkout@v2
  - name: Set up Python 3.x
    uses: actions/setup-python@v2
    with:
      python-version: '3.7'    
   
  - uses: actions/cache@master
    id: cache
    with:
      path: ${{ env.pythonLocation}}/lib/python3.7/site-packages/*
      key: ${{ runner.os }}-pip-${{ hashFiles('requirements.txt') }}
      restore-keys: |
        ${{ runner.os }}-pip-
        ${{ runner.os }}-
        
  - name: Install Dependencies
    if: steps.cache.outputs.cache-hit != 'true'
    run: |
      echo "Installing dependencies and caching them."
      pip install -r requirements.txt
 
  - name: ls show env.pythonLocation
    run: ls -R ${{ env.pythonLocation}}
 
  - name: show module
    run: pip show pandas
  
  - name: Train model and log wandb metric
    run:  echo "Finished run."</code></pre>
